Feat/save wakeup time

zackerms opened the pull request:
## 実装の背景・目的

利用者が「おはよう」と挨拶したときに、自動的に起床時間が記録されるようにした。
これにより、利用者が起床時間を記録する手間を省くことができる。

## やったこと

- 「おはよう」と挨拶すると、起床時間を記録する
-  記録が完了すると、PUSHでお知らせ
    - 記録に時間がかかる場合を想定して、あえてPUSH通知にしています
- LINEユーザーから、このAPI用のユーザーを自動的に作成 
- １日に複数回、起床時間を記録することはできません
  - 既にその日の起床時間が記録されているかどうかを確認する`User#today_wakeup_saved?`に、単体テストを作成しています。

## キャプチャ

- １日に一回だけ記録がされます
![スクリーンショット 2021-11-19 17 45 40](https://user-images.githubusercontent.com/55840281/142592956-aaddc3ca-7a0a-42e9-a393-43bc29cdc9df.png)

## 補足

- 利用者が何日連続で早起きしたかを報告する機能を付けていきたいです！
- 利用者を褒めて、目に見える形で成果を提示することで、生活習慣が良い方向に向くようにしたいです。　

zackerms committed:
feat(model): create user from LINE

LINEユーザーからUserを作成するメソッドを作成した。
アプリケーションがLINEに依存しないようにするため、Userを介して様々な操作ができるようにする。

zackerms committed:
feat(app): configure db timezone

DBの時間が日本時間で取得できるように
設定を追加した。

zackerms committed:
feat(controller): save wakeup time

利用者が「おはよう」と挨拶したときに、
起床時間が自動的に記録されるように機能を追加した。

zackerms committed:
feat(model, controller): save wakeup time only once per day

zackerms committed:
refactor(model): use create to save model

モデル作成時により短く書けるように、createメソッドを利用するように変更した。

zackerms committed:
build(db): modified database configuration required for heroku

egurinko commented on app/controllers/webhook_controller.rb:
```diff
@@ -168,4 +170,41 @@ def good_morning_response(client, event)
+    user_id = event["source"]["userId"]
+
+    # ユーザーを検索（登録されていなければ登録する）
+    user = User.find_by(line_user_id: user_id)
```
rails には find_or_create_by! という最強の method があるのです 👀 

ref: https://railsdoc.com/page/find_or_create_by

zackerms replied:
そんな便利なメソッドがあるんですね！
それを使えば、User#create_from_line_userが要らなくなりそうです。

egurinko commented on app/controllers/webhook_controller.rb:
```diff
@@ -168,4 +170,41 @@ def good_morning_response(client, event)
+    end
+
+    # 既に同じ日の起床時間が記録されていない場合は何もしない
+    if user.today_wakeup_saved?
+      return
+    end
```
User モデルの中にあるメソッドが wakeup の保存に問題ないかバリデーションするのは責務的に若干怪しくて、wakeup のことは Wakeup モデルに書いてあげると責務的にはさらに良くなりそうです（努力目標）。

```rb
def save_wakeup_time
    return if event["source"]["type"] != "user"
    
    user_id = event["source"]["userId"]
    user = User.find_or_create_by!(line_user_id: user_id)

    user.wake_ups.create!(wakeup_at: Time.zone.now)
    client.push_message(user_id, create_text_object("起きた時間を記録したよ！"))
    
    rescue => e
        Rails.logger.error("[WebhookController] Error while saving wakeup time: #{e}")
        client.push_message(user_id, create_text_object("起床時間を記録することができなかったよ。\n明日、もう一度、声をかけてみて！"))
    end
end

// models/wakup.rb
class Wakeup < AR
    validate :today_wakeup_saved

    def today_wakeup_saved
        # validation のロジックをここに書いてあげる ref: https://qiita.com/h1kita/items/772b81a1cc066e67930e#%E3%82%AB%E3%82%B9%E3%82%BF%E3%83%A0%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89
    end
end
```

zackerms replied:
その人（User）の起床記録があるかどうかというのが知りたいと思ったときに、
Userにあるのが自然かなと思いました！

でも、確かに、`Wakeup`関連のロジックが`User`にあると、ほかのロジックが混じったときに`User`がパンクしそうですね。

egurinko replied:
あくまで Wakeup に関することは Wakeup に任せてあげる方がコードが自然になりますね 👀 
とはいえ、カスタムバリデーションまで手を出すとなかなかしんどい気がするので、find_or_create_by! だけある程度修正すると良いかも

egurinko commented on app/controllers/webhook_controller.rb:
```diff
@@ -168,4 +170,41 @@ def good_morning_response(client, event)
+
+    # 起床時間を記録する
+    begin
+      user.save_wakeup(Time.now)
```
rails の難しいところの 1 つなんですが、has__many の関連を正しく書けていれば `user.wakeups.create!` のように書くことができます。ちなみに、has_one のリレーションだと、`user.create_wakeup()` みたいになります。

また、めっちゃ細かいですが save_wakeup の中で create! メソッドを呼んで例外をあげるようにしているので、save_wakeup も `save_wakeup!` にしてあげたほうが良きですね

zackerms replied:
`user.wakeups.create!(wakeup_at: Time.now)`で起床時間を記録することができました。
`User`から`Wakeup`関連のメソッドを無くすことができるので、
こちらの修正で行きたいと思います！

egurinko commented on app/models/user.rb:
```diff
@@ -1,3 +1,29 @@
   has_many :wakeups
+
+  # LINEのuser idからユーザーを作成する
+  def self.create_from_line_user(user_id)
```
モデルにメソッド定義して controller から使えているのとても良いです 👌 
また、クラスメソッドとインスタンスメソッドを使い分けられているのもとても良きです

egurinko commented on config/database.yml:
```diff
@@ -80,6 +80,4 @@ test:
-  database: ruby-getting-started_production
-  username: ruby-getting-started
-  password: <%= ENV['RUBY-GETTING-STARTED_DATABASE_PASSWORD'] %>
+  password: <%= ENV['DATABASE_URL'] %>
```
[ask]
変更が混ざっちゃってそう？？

zackerms replied:
この機能からDBアクセスが必要になるので、Herokuの設定を書き換える必要があるなと思いました。
ただ、その変更は、確かに分けたほうが良かったかもしれないです！！

egurinko replied:
そうですね、トピック的には分けられるとスムーズっすね 🙆‍♂️ 

ちなみに、password にデータベース URL 突っ込むと疎通できることを初めて知りました。
本番環境だと、URL に含まれている mysql の username, password, port, host などをしっかり分けて database.yml から読み込むことが多いですね 👀  こうして分離しておくと、各種変更に柔軟に対応しやすくなります。

```
database: <%= ENV["DATABASE_DATABASE"] %>
username: <%= ENV["DATABASE_USERNAME"] %>
password: <%= ENV["DATABASE_PASSWORD"] %>
host: <%= ENV["DATABASE_HOST"] %>
port: <%= ENV["DATABASE_PORT"] %>
```

zackerms committed:
refactor(user): removed create_from_line_user method

`find_or_create_by!`メソッドを利用することで、
同様の処理を書くことができるため、Userが持っていた
LINEユーザーからUserを作るメソッドを削除した。

zackerms committed:
refactor(user): removed save wakeup method

UserとWakeupとの関連により、
起床時間を利用者と紐付けるActiveRecordのメソッドを
利用することができるので、全く同じ動きをしていたメソッドを削除した。

egurinko commented on app/models/user.rb:
```diff
@@ -1,3 +1,14 @@
+
+  # 今日の起床時間が記録されているかどうか
+  def today_wakeup_saved?
+    last_wakeup = wakeups.order(wakeup_at: :desc).first
```
wakeup_at が今日のものを取得するメソッドは以下のようにシンプルにも書けますね 👀 

```rb
class Wakeup < ApplicationRecord
    scope :today_wakeups, -> { where(created_at: Time.zone.now.all_day) }
end

class User < ApplicationRecord
    def today_wakeup_saved?
        Wakeup.today_wakeups.count > 0
    end
end
```

zackerms replied:
`scope`という機能があるのですね！
確かに、これなら簡潔にかけますね

ただ、使い方がまだ分からないので、一旦このままにしておきたいと思います。
（次からは使えるように、勉強します！）

egurinko reviewed (approved):
実装自体は問題なさそうです 👍

masakizk commented:
ありがとうございます！
では、マージします。
